### ISSUE TITLE
Show repeats a second time!

### ISSUE THREAD

**Issue by CrimsonGhost138** [LibreTime/libretime#1170]
We have the problem of repeating shows. Scheduled a 1 hour long mp3 file in the calendar manually. After that 1 hour show, there should be something else playing, but it repeats the same mp3 file a second time. After second play, it goes on, with a little delay, with the show planned for that time.

I am afraid i can´t really see a clear line, in how to replicate that behaviour. We have a 24/7 program, and MOST of the time, everything works as expected, but it happens mostly when:

.) Schedule has no autoloading Playlist and only 1 or 2 files in it
.) Block is around 1 hour and the mp3 file is almost exactly 1 hour long

**LibreTime version:**
0f7fbf1d5

**Comment by paddatrapper** [comment]
You mention a block, is that what is scheduled in the show?

**Comment by CrimsonGhost138** [comment]
No, i am sorry, it´s just a show created in the calendar, no smart blocks, no playlist, ony 1 hour mp3 file, then there is the next "show". 
I am trying to analyse the log files but can´t find a clue yet

In the GUI itself, i can see the correct files of the NEXT show playing (while the mp3 runs a second time), but it´s not hearable, neither can you see it in the "Playout History". 

So something goes really wrong here...

**Comment by Robbt** [comment]
I have had this happen too. Do you have any way to access your various logs under /var/log/airtime/pypo and /var/log/airtime/pypo-liquidsoap  to see what errors might be occurring.

If we could figure out how to recreate this it would be most helpful. I usually find an error relating to liquidsoap or pypo not being able to remove a file because it is in use.

**Comment by CrimsonGhost138** [comment]
Yes i think i´ve seen something like that, too!

Here is the log file, the error has occured with a show, broadcasted today, 26.1., at 09:00-10:00, and repeated itself at 10:00 until the mp3 file finished a second time. Then, it played back the correct audio for that show. So it seems like it repeated the mp3 file 2 times, because something happened when it finished before?

http://grenzwelle.kumt.org/upload/pypo.log

**Comment by Robbt** [comment]
Yeah, it is the same problem I've experienced. I just don't know what causes it or how to fix it.

It happens seemingly randomly at my station but here is the line that indicates the error 


2021-01-26 10:01:57,972 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:09:58,432 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:17:59,152 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:25:59,753 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:34:00,366 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:42:00,953 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!
2021-01-26 10:50:01,627 [pypofetch] [INFO ]  File '/var/tmp/airtime/pypo/cache/scheduler/6315.mp3' not removed. Still busy!

And here is the pastebin with everything including that line but it tends to be hard to parse the new schedule received because its such a dense list of files to determine if there is an error there. 

https://pastebin.com/8Eycwknz

This is definitely already an issue somewhere but maybe we can figure out how to override removal of files even if they're busy and just tell liquidsoap to to the scheduler.

**Comment by CrimsonGhost138** [comment]
OK it happened again! Now i am about to see some co-incidence: All the files have been exported and converted with Audacity. I am investigating this further and keep you updated!!!

**Comment by Robbt** [comment]
I'm still running into this problem intermittently. My station is running an older version of LibreTime but I'm not sure we ever fully resolved this problem. 

Essentially the fix is to restart pypo-liquidsoap and pypo manually when the schedule falls behind.

I had the idea that perhaps LibreTime should restart pypo-liquidsoap when it can't delete a file (that should be removed) because that means that there is a disconnect between what should be playing and what is playing. 

Just a thought, I haven't really had much time to get into Libre

**Comment by Robbt** [comment]
I think the priority here should be to probably to not ignore "not removed. Still busy!" - but to for this to trigger some kind of restart of the whole system.

I can usually resolve this issue by restarting pypo and liquidsoap but then it'll start playing the show as if it had already been playing.

So many the solution here is to override. I'll see if any of the relevant code has changed since this issue was first happening and then maybe I'll try write something that responds differently, rather than writing a log it'll attempt to rectify the problem.

**Comment by Robbt** [comment]
So I don't think that this commit https://github.com/libretime/libretime/commit/05b988fc6ab16d5676770022ac2962bfd483dbc4 caused the problem which is libretime playing older files when it should be playing different files but it is the error that usually manifests in the logs. 

Here are the lines of code that I think we could utilize to detect that LibreTime is out of sync and initiate some kind of reset to liquidsoap to fix itself.

https://github.com/libretime/libretime/blame/5a7e4db23e0abeb39a917155fad56928316c4fa6/playout/libretime_playout/player/fetch.py#L245-L262

I'd rather discover where the bug is that causes the files to be played longer than they should but I'm perfectly happy to throw a band-aid on this bug as long as it doesn't cause other issues like dead air.

The simplest thing to try would be to restore the previous functionality of deleting the file even if liquidsoap is playing it but still throwing out a log. The question that could be answered by manually doing this on a test server is what happens in this case. It could trigger dead air which would be worse but if playout & liquidsoap scripts figure this out and start playing the next file in the queue then this is preferrable at least for my station. Another solution and this has been what I've done when DJs call me after they realize their show isn't working is to stop liquidsoap and stop playout and then restart them and it fixes itself. 

I don't know if this will fix itself with just restarting one of those scripts, having a script restart itself when it reaches an error state seems kind of like a janky fix that could cause other problems like endless reboots. Especially if we restarted both of them. I also wouldn't know how to do this from inside of the script itself.

The other and probably eventually best thing to do would be to trace things up the stack and determine what actually causes this error, and really analyze the back and forth between the playout and liquidsoap-python script and the actual liquidsoap script and see if we can figure out what causes the error. 

It is also possible that this is a liquidsoap bug as #1488 ran into this error and it was resolved by newer software and updated liquidsoap and ffmpeg.

One thing I could do is upgrade LibreTime and see if this is resolved but that involves essentially building a brand new server and transferring the files.

Since there haven't been any issues opened up for this in over 2 years it's quite possible that this bug was related to filesize problems and improved checks have resolved it. 

I'll also need to test the conditions on #232 and see if this can be replicated with the newest software and if not then we can close these two longstanding bugs.
